src/util/range/numeric_range.c

```diff
@@ -0,0 +1,125 @@
+/*
+* Copyright 2018-2019 Redis Labs Ltd. and Contributors
+*
+* This file is available under the Redis Labs Source Available License Agreement
+*/
+
+#include "numeric_range.h"
+#include "../rmalloc.h"
+#include "../../ast/ast_shared.h"
+#include <math.h>
+#include <assert.h>
+
+NumericRange *NumericRange_New(void) {
+	NumericRange *range = rm_malloc(sizeof(NumericRange));
+	range->valid = true;
+	range->max = INFINITY;
+	range->min = -INFINITY;
+	range->include_min = false;
+	range->include_max = false;
+	return range;
+}
+
+bool NumericRange_IsValid(const NumericRange *range) {
+	if(range->include_min && range->include_max) {
+		// X >= y AND X <= z
+		return (range->min <= range->max);
+	} else {
+		// X >= y AND X < z
+		// X > y AND X <= z
+		// X > y AND X < z
+		return (range->min < range->max);
+	}
+}
+
+bool NumericRange_ContainsValue(const NumericRange *range, double v) {
+	if(!range->valid) return false;
+
+	// Make sure v is <= max.
+	if(range->include_max) {
+		if(v > range->max) return false;
+	} else {
+		if(v >= range->max) return false;
+	}
+
+	// Make sure v >= min.
+	if(range->include_min) {
+		if(v < range->min) return false;
+	} else {
+		if(v <= range->min) return false;
+	}
+
+	return true;
+}
+
+void NumericRange_TightenRange(NumericRange *range, int op, double v) {
+	if(!range->valid) return;
+
+	switch(op) {
+	case OP_LT:    // <
+		if(range->max >= v) {
+			range->include_max = false;
+			range->max = v;
+		}
+		break;
+	case OP_LE:    // <=
+		if(range->max > v) {
+			range->include_max = true;
+			range->max = v;
+		}
+		break;
+	case OP_GT:    // >
+		if(range->min <= v) {
+			range->include_min = false;
+			range->min = v;
+		}
+		break;
+	case OP_GE:    // >=
+		if(range->min < v) {
+			range->include_min = true;
+			range->min = v;
+		}
+		break;
+	case OP_EQUAL:  // =
+		// Make sure v is within range.
+		if(!NumericRange_ContainsValue(range, v)) {
+			range->valid = true;
+			return;
+		}
+
+		range->include_min = true;
+		range->include_max = true;
+		range->min = v;
+		range->max = v;
+		break;
+	}
```

Comment:
Needs some kind of handling for OP_NEQUAL, since right now the filter is just discarded:
`MATCH (a:person) WHERE a.age <> 30 RETURN a`

Reply:
See my comment on string-range.

Reply:
My testing showed that inequality filters were not being converted properly. The above query forms an index scan, but results where `age == 30` were included in the result set.

Reply:
Added logic to skip index utilization for filters using the NE operation